blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,958 @@
+					"type": "uint256"
+				},
+				{
+					"indexed": true,
+					"internalType": "address",
+					"name": "recipient",
+					"type": "address"
+				}
+			],
+			"name": "Withdrawal",
+			"type": "event"
+		}
+	]`
+
+	// bucket related namespace in db
+	_liquidStakingBucketInfoNS = "lsbInfo"
+	_liquidStakingBucketTypeNS = "lsbType"
+	_liquidStakingNS           = "lsns"
+)
+
+type (
+	// LiquidStakingIndexer is the interface of liquid staking indexer
+	LiquidStakingIndexer interface {
+		blockdao.BlockIndexer
+
+		CandidateVotes(candidate address.Address) *big.Int
+		Buckets() ([]*Bucket, error)
+		Bucket(id uint64) (*Bucket, error)
+		BucketsByIndices(indices []uint64) ([]*Bucket, error)
+		TotalBucketCount() uint64
+		ActiveBucketTypes() (map[uint64]*BucketType, error)
+	}
+
+	// liquidStakingIndexer is the implementation of LiquidStakingIndexer
+	// Main functions:
+	// 		1. handle liquid staking contract events when new block comes to generate index data
+	// 		2. provide query interface for liquid staking index data
+	// Generate index data flow:
+	// 		block comes -> new dirty cache -> handle contract events -> update dirty cache -> merge dirty to clean cache
+	// Main Object:
+	// 		kvstore: persistent storage, used to initialize index cache at startup
+	// 		cache: in-memory index for clean data, used to query index data
+	//      dirty: the cache to update during event processing, will be merged to clean cache after all events are processed. If errors occur during event processing, dirty cache will be discarded.
+	liquidStakingIndexer struct {
+		kvstore       db.KVStore                      // persistent storage
+		cache         *liquidStakingCacheThreadSafety // in-memory index for clean data
+		blockInterval time.Duration
+	}
+)
+
+var (
+	_liquidStakingInterface           abi.ABI
+	_liquidStakingHeightKey           = []byte("lsHeight")
+	_liquidStakingTotalBucketCountKey = []byte("lsTotalBucketCount")
+
+	errBucketTypeNotExist = errors.New("bucket type does not exist")
+
+	// ErrBucketInfoNotExist is the error when bucket does not exist
+	ErrBucketInfoNotExist = errors.New("bucket info does not exist")
+)
+
+func init() {
+	var err error
+	_liquidStakingInterface, err = abi.JSON(strings.NewReader(LiquidStakingContractABI))
+	if err != nil {
+		panic(err)
+	}
+}
+
+// NewLiquidStakingIndexer creates a new liquid staking indexer
+func NewLiquidStakingIndexer(kvStore db.KVStore, blockInterval time.Duration) LiquidStakingIndexer {
+	return &liquidStakingIndexer{
+		blockInterval: blockInterval,
+		kvstore:       kvStore,
+		cache:         newLiquidStakingCache(),
+	}
+}
+
+// Start starts the indexer
+func (s *liquidStakingIndexer) Start(ctx context.Context) error {
+	if err := s.kvstore.Start(ctx); err != nil {
+		return err
+	}
+	return s.loadCache()
+}
+
+// Stop stops the indexer
+func (s *liquidStakingIndexer) Stop(ctx context.Context) error {
+	if err := s.kvstore.Stop(ctx); err != nil {
+		return err
+	}
+	return nil
```

Comment:
should the cache be reset here?

Reply:
strictly speaking, it is recommended to reset the cache upon `Stop` to ensure proper functioning when it is re-`Start`, even though it is not actually utilized in that manner.